internal/provider/resource_access_model.go

```diff
@@ -5,186 +5,314 @@ import (
+func (a *accessModelResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "update")
+		return
+	}
+
+	var plan, state accessModelResourceModel
+
+	// Get the Terraform state from the request into the plan
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	var diags diag.Diagnostics
 	anyChange := false
 
 	// items that could be changed
 	var newAccess string
-	var newUsersList []string
 	var missingUserList []string
 	var addedUserList []string
 
-	var err error
+	// Get the users that are in the planned state
+	planUsers := convertRawUsers(plan.Users.Elements())
 
-	if d.HasChange("users") {
+	// Check if the users has changed
+	if !plan.Users.Equal(state.Users) {
 		anyChange = true
-		oldUsers, newUsers := d.GetChange("users")
-		oldUsersInterface := oldUsers.([]interface{})
-		oldUsersList := make([]string, len(oldUsersInterface))
-		for i, v := range oldUsersInterface {
-			oldUsersList[i] = v.(string)
-		}
-		newUsersInterface := newUsers.([]interface{})
-		newUsersList = make([]string, len(newUsersInterface))
-		for i, v := range newUsersInterface {
-			newUsersList[i] = v.(string)
-		}
-		missingUserList = getMissingUsers(oldUsersList, newUsersList)
-		addedUserList = getAddedUsers(oldUsersList, newUsersList)
+
+		// Get the users that are in the current state
+		stateUsers := convertRawUsers(state.Users.Elements())
+
+		missingUserList = getMissingUsers(stateUsers, planUsers)
+		addedUserList = getAddedUsers(stateUsers, planUsers)
+	}
+
+	// Check if access has changed
+	if !plan.Access.Equal(state.Access) {
+		anyChange = true
+		newAccess = plan.Access.ValueString()
 	}
 
 	if !anyChange {
-		return diags
+		tflog.Trace(ctx, "Update is returning without any changes.")
+		return
+	}
+
+	modelName, oldAccess, _ := retrieveAccessModelDataFromID(state.ID, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	err = client.Models.UpdateAccessModel(juju.UpdateAccessModelInput{
-		Model:  d.Id(),
-		Grant:  addedUserList,
-		Revoke: missingUserList,
-		Access: newAccess,
+	err := a.client.Models.UpdateAccessModel(juju.UpdateAccessModelInput{
+		ModelName: modelName,
+		OldAccess: oldAccess,
+		Grant:     addedUserList,
+		Revoke:    missingUserList,
+		Access:    newAccess,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update access model resource, got error: %s", err))
 	}
+	tflog.Trace(ctx, fmt.Sprintf("updated access model resource for model %q", modelName))
+
+	plan.ID = types.StringValue(newAccessModelIDFrom(modelName, newAccess, planUsers))
```

Comment:
the ID is incorrectly written if only the users are modified, but no the access type.
```
juju_user.second: Creation complete after 0s [id=user:cherry]
juju_access_model.this: Modifying... [id=usertest:write:coke]
juju_access_model.this: Modifications complete after 0s [id=usertest::coke,cherry]
```